### PR TITLE
Make location picker pin visible and colors match theme

### PR DESCRIPTION
**Pull request by bencodeorg**
Match background and icon in location picker to theme, and improve visibility in high contrast mode.

Theme | Before      | After |
| ----------- | ----------- | ----------- | 
| High Contrast | image | image |
| Modern   | image        | image |

## Links

- Jira: [link](https://codedotorg.atlassian.net/browse/SL-644)

## Testing story

Tested manually on a poetry project level across themes available in the context menu.

### CODE REVIEW

**Comment by mikeharv** [comment]
I pulled and tested these changes and confirmed that everything works and responds as expected. I checked the block shown in the PR description across themes. I also checked standard location picker from Sprite Lab by adding it to a Poetry level. The field looks to get updated on theme change and also renders correctly on page load.

image
image
image

The contrast is less great when using the Zelos renderer, currently only possible with an experiment. I'm guessing it uses different logic to choose its secondary colors. Given that we're not planning to actually use Zelos, I think this is fine, but it's good to see that it at least renders as expected.
image

I think the code changes here seem very reasonable.

File: apps/src/blockly/addons/cdoFieldButton.js [hunk 1]
```diff
@@ -68,4 +69,14 @@ export default class CdoFieldButton extends GoogleBlockly.Field {
       this.buttonElement_.setAttribute('width', this.size_.width + 8);
     }
   }
+
+  positionBorderRect_() {
+    super.positionBorderRect_();
+    if (this.buttonElement_) {
```
Comment (maribethb) [inline apps/src/blockly/addons/cdoFieldButton.js:100]: I think you want to do some of this work in an `applyColour` [method](https://developers.google.com/blockly/guides/create-custom-blocks/fields/customizing-fields/creating#matching_block_colours). The block will automatically call this method when the block is rendering and it's where internally we do some related kinds of rendering updates. Specifically if you need to update the colors of a field based on its parent block, that's what this method is for. So the last two lines can go in `applyColour`.


As for setting the height and width, that's where I'm a little confused. You have this `updateWidth_` method defined, and it says it calls `super`, but `GoogleBlockly.Field` doesn't have a method called `updateWidth_`. The cdo fork does, but I don't see how that could be relevant here since we inherit directly from `GoogleBlockly.Field`. My guess is this `updateWidth_` method is totally unused and if you tried to call it, you would get an error because `super.updateWidth_()` does not exist.

Also, in that function, you set the width differently than how you're setting it in this new code, and that's confusing to me, but maybe the calculation is just out of date since the `updateWidth_` calculation is not used anywhere.

Instead, to update the size of the field, you can do that work in `updateSize_`. You will still need to call `super.updateSize_` as well. Then you can set the width and height attributes in that method.

TLDR: Update the height and width properties in `updateSize_` and update the fill colors in `applyColours` (note the uk spelling of colour)

Comment (maribethb) [reply apps/src/blockly/addons/cdoFieldButton.js:100]: For more info on why: 

The current code works because `updateSize_` (which is called when the field value changes or when the block is re-rendered e.g. when the renderer/theme changes) calls `positionBorderRect`. So you've found a method that is called at the appropriate time and tacked on some unrelated work to it, but it would be better to put that work in a more appropriate spot.


`applyColour` is called only when the color of the field might need to change, e.g. parent block changed, `changeColour` was called on a block directly, or the renderer changed. That is a better time to update the fill colors because you don't need to update them every time the value of the field changes like you would be doing with the current code.

Comment (bencodeorg) [reply apps/src/blockly/addons/cdoFieldButton.js:100]: This feels much better, thank you Maribeth! I also confirmed calling `super.updateWidth_()` breaks, which makes sense -- fully removed that function.

Curious where I should have looked to see the `applyColour` info in the docs (seems quite obvious based on the link you shared 🤦 ) -- I didn't see any reference to it in the `Field` docs?

**Review by mikeharv — APPROVED** [review]

File: apps/src/blockly/addons/cdoFieldButton.js [hunk 2]
```diff
@@ -62,10 +62,22 @@ export default class CdoFieldButton extends GoogleBlockly.Field {
     this.buttonHandler_(this.setValue.bind(this));
   }
 
-  updateWidth_() {
-    super.updateWidth_();
+  updateSize_() {
```
Comment (fisher-alice) [inline apps/src/blockly/addons/cdoFieldButton.js:83]: What do you think about add a comment when we're overriding functions vs adding new functions to this class?
Also, this is not part of your changes, but I noticed that there is already a `getValue` function defined in `field.ts` so we could remove it from this class.

File: apps/src/blockly/addons/cdoFieldButton.js [hunk 3]
```diff
@@ -4,12 +4,11 @@ const CORNER_RADIUS = 3;
 const INNER_HEIGHT = 16;
 
 export default class CdoFieldButton extends GoogleBlockly.Field {
-  constructor(title, opt_buttonHandler, opt_color, opt_changeHandler) {
+  constructor(title, opt_buttonHandler, _, opt_changeHandler) {
```
Comment (fisher-alice) [inline apps/src/blockly/addons/cdoFieldButton.js:9]: A comment here could be helpful to understand why we don't need the third parameter - it seems like we're juggling between CDO vs mainline blockly.

**Review by fisher-alice — APPROVED** [review]
Nice work, Ben!

File: apps/src/blockly/addons/cdoFieldButton.js (hunk 2 above)
Comment (bencodeorg) [reply apps/src/blockly/addons/cdoFieldButton.js:83]: Looking at the google blockly field class, I think these are all overrides technically? Looks like their getValue doesn't guarantee returning a string (if my typescript comprehension is correct?), might want to use getText instead. And I see at least the `setValue` override is not recommended (maybe others as well 😬 ), [link](https://github.com/google/blockly/blob/fa9cb579ae119749ee14277518374060bc555c3e/core/field.ts#L983-L990)

File: apps/src/blockly/addons/cdoFieldButton.js (hunk 1 above)
Comment (maribethb) [reply apps/src/blockly/addons/cdoFieldButton.js:100]: How to find stuff in the docs: I always look in the Guides section first, and there are several pages about fields. The bulk of the info about the APIs are on the [Creating a Custom Field](https://developers.google.com/blockly/guides/create-custom-blocks/fields/customizing-fields/creating) page.

If I don't see anything in the Guides section, I'll check Reference then, but this is a bit more spotty as it's auto-generated. I noticed the `applyColour` method is not there so I'll fix that for next release.

Comment (bencodeorg) [reply apps/src/blockly/addons/cdoFieldButton.js:100]: Cool I always go Reference first, good to know I should start with Guides!

File: apps/src/blockly/addons/cdoFieldButton.js (hunk 2 above)
Comment (fisher-alice) [reply apps/src/blockly/addons/cdoFieldButton.js:83]: Interesting! It seems like the value stored by the field is ensured to be a certain type by the validation done in `setValue` And yeah, I see the recommendation in the `setValue` comment to override `doClassValidation_` and `doValueUpdate_`. Maybe in a later refactor we override `doValueUpdate_` instead of `setValue`?
